utils/src/migration.rs

```diff
@@ -1,33 +1,219 @@
+pub enum ApiVersion<CR> {
+    /// alpha version e.g. v1alpha1 => Alpha(1, 1)
+    Alpha(u32, u32, PhantomData<CR>),
+    /// beta version e.g. v2beta3 => Beta(2, 3)
+    Beta(u32, u32, PhantomData<CR>),
+    /// stable version e.g v3 => Stable(3)
+    Stable(u32, PhantomData<CR>),
+}
+
+impl<CR> ApiVersion<CR> {
+    /// Create alpha version
+    #[must_use]
+    #[inline]
+    pub fn alpha(main: u32, sub: u32) -> Self {
+        Self::Alpha(main, sub, PhantomData)
+    }
+
+    /// Create beta version
+    #[must_use]
+    #[inline]
+    pub fn beta(main: u32, sub: u32) -> Self {
+        Self::Beta(main, sub, PhantomData)
+    }
+
+    /// Create beta version
+    #[must_use]
+    #[inline]
+    pub fn stable(main: u32) -> Self {
+        Self::Stable(main, PhantomData)
+    }
+
+    /// return the main version
+    fn main_version(&self) -> u32 {
+        match *self {
+            ApiVersion::Stable(main, _)
+            | ApiVersion::Beta(main, _, _)
+            | ApiVersion::Alpha(main, _, _) => main,
+        }
+    }
+
+    /// return the sub version
+    fn sub_version(&self) -> u32 {
+        match *self {
+            ApiVersion::Stable(_, _) => 0,
+            ApiVersion::Beta(_, sub, _) | ApiVersion::Alpha(_, sub, _) => sub,
+        }
+    }
+
+    /// return the numeric label order for comparing
+    fn label_order(&self) -> u32 {
+        match *self {
+            ApiVersion::Alpha(_, _, _) => 0,
+            ApiVersion::Beta(_, _, _) => 1,
+            ApiVersion::Stable(_, _) => 2,
+        }
+    }
+}
+
+impl<CR> FromStr for ApiVersion<CR> {
+    type Err = anyhow::Error;
+
+    #[inline]
+    #[allow(clippy::indexing_slicing)] // it is obvious that a string start with 'v' has the index 0
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.starts_with('v') {
+            return Err(anyhow!(
+                "invalid api version format, version does not start with 'v'"
+            ));
+        }
+        if s.contains("alpha") {
+            let parts: Vec<_> = s[1..].split("alpha").collect();
+            if parts.len() != 2 {
+                return Err(anyhow!("invalid api version format"));
+            }
+            let main_ver = parts[0].parse()?;
+            if parts[1].is_empty() {
+                return Ok(Self::Alpha(main_ver, 0, PhantomData));
+            }
+            let sub_ver = parts[1].parse()?;
+            return Ok(Self::Alpha(main_ver, sub_ver, PhantomData));
+        }
+        if s.contains("beta") {
+            let parts: Vec<_> = s[1..].split("beta").collect();
+            if parts.len() != 2 {
+                return Err(anyhow!("invalid api version format"));
+            }
+            let main_ver = parts[0].parse()?;
+            if parts[1].is_empty() {
+                return Ok(Self::Beta(main_ver, 0, PhantomData));
+            }
+            let sub_ver = parts[1].parse()?;
+            return Ok(Self::Beta(main_ver, sub_ver, PhantomData));
+        }
```

Comment:
This two parts are very similar, should we use a macro or function to reduce duplication